src/iceberg/expression/aggregate.cc

```diff
@@ -0,0 +1,289 @@
+ */
+
+#include "iceberg/expression/aggregate.h"
+
+#include <format>
+#include <optional>
+
+#include "iceberg/exception.h"
+#include "iceberg/expression/binder.h"
+#include "iceberg/expression/expression.h"
+#include "iceberg/row/struct_like.h"
+#include "iceberg/type.h"
+#include "iceberg/util/checked_cast.h"
+#include "iceberg/util/macros.h"
+
+namespace iceberg {
+
+namespace {
+
+std::string OperationToPrefix(Expression::Operation op) {
+  switch (op) {
+    case Expression::Operation::kMax:
+      return "max";
+    case Expression::Operation::kMin:
+      return "min";
+    case Expression::Operation::kCount:
+    case Expression::Operation::kCountStar:
+      return "count";
+    default:
+      break;
+  }
+  return "aggregate";
+}
+
+Result<std::shared_ptr<PrimitiveType>> GetPrimitiveType(const BoundTerm& term) {
+  auto primitive = std::dynamic_pointer_cast<PrimitiveType>(term.type());
+  if (primitive == nullptr) {
+    return InvalidExpression("Aggregate requires primitive type, got {}",
+                             term.type()->ToString());
+  }
+  return primitive;
+}
+
+}  // namespace
+
+CountAggregate::CountAggregate(Expression::Operation op, Mode mode,
+                               std::shared_ptr<UnboundTerm<BoundReference>> term,
+                               std::shared_ptr<NamedReference> reference)
+    : UnboundAggregate(op),
+      mode_(mode),
+      term_(std::move(term)),
+      reference_(std::move(reference)) {}
+
+Result<std::unique_ptr<CountAggregate>> CountAggregate::Count(
+    std::shared_ptr<UnboundTerm<BoundReference>> term) {
+  auto ref = term->reference();
+  return std::unique_ptr<CountAggregate>(new CountAggregate(
+      Expression::Operation::kCount, Mode::kNonNull, std::move(term), std::move(ref)));
+}
+
+Result<std::unique_ptr<CountAggregate>> CountAggregate::CountNull(
+    std::shared_ptr<UnboundTerm<BoundReference>> term) {
+  auto ref = term->reference();
+  return std::unique_ptr<CountAggregate>(new CountAggregate(
+      Expression::Operation::kCount, Mode::kNull, std::move(term), std::move(ref)));
+}
+
+std::unique_ptr<CountAggregate> CountAggregate::CountStar() {
+  return std::unique_ptr<CountAggregate>(new CountAggregate(
+      Expression::Operation::kCountStar, Mode::kStar, nullptr, nullptr));
+}
+
+std::string CountAggregate::ToString() const {
+  if (mode_ == Mode::kStar) {
+    return "count(*)";
+  }
+  ICEBERG_DCHECK(reference_ != nullptr, "Count aggregate should have reference");
+  switch (mode_) {
+    case Mode::kNull:
+      return std::format("count_null({})", reference_->name());
+    case Mode::kNonNull:
+      return std::format("count({})", reference_->name());
+    case Mode::kStar:
+      break;
+  }
+  std::unreachable();
+}
+
+Result<std::shared_ptr<Expression>> CountAggregate::Bind(const Schema& schema,
+                                                         bool case_sensitive) const {
+  std::shared_ptr<BoundTerm> bound_term;
+  if (term_ != nullptr) {
```

Comment:
Please write the code like `if (term_ != nullptr) [[unlikely]] {` to help branch prediction.